compiler/rustc_mir_build/src/build/mod.rs

```diff
@@ -403,6 +403,43 @@ enum NeedsTemporary {
 #[must_use = "if you don't use one of these results, you're leaving a dangling edge"]
 struct BlockAnd<T>(BasicBlock, T);
 
+impl<T> BlockAnd<T> {
+    /// Unpacks `BlockAnd<T>` by writing the block to the given reference,
+    /// and then returning the value.
+    #[must_use]
+    fn unpack(self, block_mut: &mut BasicBlock) -> T {
+        let Self(block, value) = self;
+        *block_mut = block;
+        value
+    }
+
+    /// Unpacks `BlockAnd<T>` into `(BasicBlock, T)`.
+    ///
+    /// This is an alternative to [`Self::unpack`] for cases where the block
+    /// variable hasn't been initialized yet, so a `&mut` reference can't be
+    /// taken.
+    #[must_use]
+    fn unpack_to_pair(self) -> (BasicBlock, T) {
+        let Self(block, value) = self;
+        (block, value)
+    }
+}
+
+impl BlockAnd<()> {
+    /// Unpacks `BlockAnd<()>` into a [`BasicBlock`].
+    #[must_use]
+    fn unpack_block(self) -> BasicBlock {
```

Comment:
The name wasn't clear to me. How about `unpack_unit`?

Reply:
I might just try `unpack_block_and_unit` instead.